### PR TITLE
PendingReadOp: Fix ledgerEntryImpl reuse problem

### PR DESCRIPTION
**Pull request by congbobo184**
Descriptions of the changes in this PR:
In bookeeper client the ledgerEntryImpl wrongly recycle cause data parsing errors.

`org.apache.bookkeeper.client.PendingReadOp` have more `LedgerEntryRequest`

for example in PendingReadOp(1) has two LedgerEntryRequest (entryId = 1, entryId = 2)
another PendingReadOp(2) has two LedgerEntryRequest (entryId = 3, entryId = 4)

1. When `LedgerEntryRequest (entryId = 1)` read fail will close all `LedgerEntryRequest (entryId = 1, entryId = 2)` in `PendingReadOp(1)`.
2. `LedgerEntryImpl` in `LedgerEntryRequest (entryId = 2)` will be recycle.
3. `LedgerEntryRequest(entryId = 3`) will reuse this `LedgerEntryImpl` in another `PendingReadOp(2)` because the `LedgerEntryImpl` in `LedgerEntryRequest (entryId = 2)` has been recycle.
4. Another `LedgerEntryRequest(entryId = 3)` readComplete will set the ByteBuf to this `LedgerEntryImpl`
5. `LedgerEntryRequest(entryId = 2)` readComplete also use this `LedgerEntryImpl` and set the ByteBuf to this `LedgerEntryImpl`.
6. Now the `LedgerEntryImpl` the entryId is 3 but the ByteBuf is entryId = 2. 

### Motivation

fix this problem

### Changes

when `LedgerEntryRequest(entryId = 2)` reponse come can't change the `LedgerEntryImpl` ByteBuf

when LedgerEntryRequest change the complete to true, the response will not change the ByteBuf. because `LedgerEntryRequest(entryId = 2)` readComplete will check complete

the same Ledger read will in the same thread.

### CODE REVIEW

**Comment by congbobo184** [comment]
rerun failure checks

**Comment by lhotari** [comment]
Thanks @congbobo184 for making this PR.

I'm just wondering that `close` should get called once as a result of the guard by `org.apache.bookkeeper.client.PendingReadOp`'s `AtomicBoolean complete` field:

https://github.com/apache/bookkeeper/blob/10825c93bccd9e170df75ab9f2cf85cb6043f649/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L619-L622

Why isn't that sufficient?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java [hunk 1]
```diff
@@ -563,6 +565,10 @@ public long getLastAddConfirmed() {
         }
     }
 
+    public static ReadContext createReadContext(int bookieIndex, BookieId to, LedgerEntryRequest entry) {
```
Comment (codelipenghui) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:568]: Don't need to be a public method? Looks only used for tests.

**Review by codelipenghui — COMMENTED** [review]
Thanks for the great work.

The change looks good to me.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java (hunk 1 above)
Comment (hangc0276) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:568]: Could we move this method into test?

**Review by hangc0276 — APPROVED** [review]
Great work! Thanks @congbobo184 
Overall looks good to me.

**Comment by hangc0276** [comment]
> Thanks @congbobo184 for making this PR.
> 
> I'm just wondering that `close` should get called once as a result of the guard by `org.apache.bookkeeper.client.PendingReadOp`'s `AtomicBoolean complete` field:
> 
> https://github.com/apache/bookkeeper/blob/10825c93bccd9e170df75ab9f2cf85cb6043f649/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L619-L622
> 
> Why isn't that sufficient?

@lhotari  It changed the  request's `complete` flag from `false` to `true`. And when the previous request's response came back, whose request has already been closed, it will be filtered by this complete flag and prevent tampering other entry's data.

**Comment by lhotari** [comment]
> > Thanks @congbobo184 for making this PR.
> > I'm just wondering that `close` should get called once as a result of the guard by `org.apache.bookkeeper.client.PendingReadOp`'s `AtomicBoolean complete` field:
> > https://github.com/apache/bookkeeper/blob/10825c93bccd9e170df75ab9f2cf85cb6043f649/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L619-L622
> > 
> > Why isn't that sufficient?
> 
> @lhotari It changed the request's `complete` flag from `false` to `true`. And when the previous request's response came back, whose request has already been closed, it will be filtered by this complete flag and prevent tampering other entry's data.

@hangc0276 That doesn't answer my question.

I meant a different thing.

It seems that LedgerEntryRequest.close can only get called via this line

https://github.com/apache/bookkeeper/blob/017307bc67431a7616861ad09927c4e3327633d1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L644

or this line

https://github.com/apache/bookkeeper/blob/7087fda4e91e9fe8974f9d36e01e275a08bf38f1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/impl/LedgerEntriesImpl.java#L59

both of those calls should only originate from the call chains starting from PendingReadOp#submitCallback:

https://github.com/apache/bookkeeper/blob/017307bc67431a7616861ad09927c4e3327633d1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L611-L650

Why isn't the `if (!complete.compareAndSet(false, true)) {` sufficient in PendingReadOp#submitCallback?

**Comment by hangc0276** [comment]
> Why isn't the `if (!complete.compareAndSet(false, true)) {` sufficient in PendingReadOp#submitCallback?

@lhotari  The `if (!complete.compareAndSet(false, true)) {` in PendingReadOp#submitCallback just change the complete flag in PendingReadOp instance. 

But the 
```
if (complete.compareAndSet(false, true)) {
                entryImpl.close();
            }
```
Change the `complete` flag in LedgerEntryRequest. They are different instances.

**Comment by lhotari** [comment]
> Change the `complete` flag in LedgerEntryRequest. They are different instances.

yes, I have noticed that there's 2 separate `complete` fields: `org.apache.bookkeeper.client.PendingReadOp#complete` and `org.apache.bookkeeper.client.PendingReadOp.LedgerEntryRequest#complete`.

https://github.com/apache/bookkeeper/blob/017307bc67431a7616861ad09927c4e3327633d1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L58-L84

That doesn't change my original question:  LedgerEntryRequest.close seems to only get called via call chains originating from PendingReadOp.submitCallback, why isn't the PendingReadOp's complete field preventing the duplicate calls to LedgerEntryRequest.close ?

**Comment by congbobo184** [comment]
rerun failure checks

**Comment by congbobo184** [comment]
rerun failure checks

**Comment by hangc0276** [comment]
> close

@lhotari  
For example
1. there is a read request for ledger 1, entryId request range is [1, 100].  When entryId [1, 10] read requests has been sent out and entryId 1 's response come back. it will close all the following ReadEntryRequests, and `LedgerEntryImpl` object will be recycled. 

2. There is another read request for ledger 2, entryId request range is [1, 2]. It will get `LedgerEntryImpl` object, which may be the ref is the previous one. For example, ReadRequest 2:1 used the same `LedgerEntryImpl` object with ReadRequest 1:2. Then send entryId 1 request to the bookkeeper server

3.  Read request 2:1 response came back, and run into `readEntryComplete` callback. It will call `entry.complete(rctx.bookieIndex, rctx.to, buffer)` to put the buffer into entryImpl's EntryBuf. And then put it into seq list to wait for complete

4. Read request 1:2 response came back. It will run into `readEntryComplete` callback. It will call `entry.complete(rctx.bookieIndex, rctx.to, buffer)` to put the buffer into entryImpl's EntryBuf.  However, the entry has been hold by read request 2:1. **This operation will put entry (1:2) bytebuf into entry (2:1) bytebuf**.  when run into `submitCallback(BKException.Code.OK);` logic, it will be filtered by the complete flag of PendingReadOP.

5. When ledger 2 's read request [1, 2] complete and all the response came back, it will return the data to client.

6. However, the entry (2:1) 's data has been tampered by step 4. This is the root cause.

**Comment by eolivelli** [comment]
@hangc0276 
in your last explanation your are not citing that there is a read error.
My understanding is that this problem may happen only when there is a read error, in RendingReadOP#submitCallback

see: 
https://github.com/apache/bookkeeper/blob/017307bc67431a7616861ad09927c4e3327633d1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L644

in the happy path it is the client application that triggers the close of the LedgerEntryImpl

is this correct ?

**Comment by hangc0276** [comment]
> @hangc0276 in your last explanation your are not citing that there is a read error. My understanding is that this problem may happen only when there is a read error, in RendingReadOP#submitCallback
> 
> see:
> 
> https://github.com/apache/bookkeeper/blob/017307bc67431a7616861ad09927c4e3327633d1/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L644
> 
> in the happy path it is the client application that triggers the close of the LedgerEntryImpl
> 
> is this correct ?

@eolivelli  Yes, that's right. I have updated the comments and added the cite.

**Comment by eolivelli** [comment]
thank you @hangc0276 for your clarification.

**Review by eolivelli — DISMISSED** [review]
LGTM

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java [hunk 2]
```diff
@@ -110,6 +110,7 @@
 
         @Override
         public void close() {
+            complete.set(true);
```
Comment (lhotari) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:113]: wouldn't it be safer to use `if (complete.compareAndSet(false, true)) {` to shield against races?

**Review by lhotari — CHANGES_REQUESTED** [review]

**Comment by lhotari** [comment]
Thanks for the explanation @hangc0276 . I now understood how the race condition happens in `org.apache.bookkeeper.client.PendingReadOp#readEntryComplete` method and it answers my question.

To prevent the problem I think that the current code needs revisiting. Why did @congbobo184 make this change: 7a1504ff8ee6140b4f63a9a6d0c1cd3485cf5887 ?
I would assume that the close method would have to be as Congbo originally wrote
```java
        public void close() {
            if (complete.compareAndSet(false, true)) {
                entryImpl.close();
            }
        }
```

instead of the current version
```java
        public void close() {
            complete.set(true);
            entryImpl.close();
        }
```

**Comment by lhotari** [comment]
I now see the reason why Congbo chose to implement it using
```java
        public void close() {
            complete.set(true);
            entryImpl.close();
        }
```

That seems to be since `entryImpl.close()` should be called when `close` is called regardless of whether the complete has been set to true.
The reason to set complete to true is to prevent using `entryImpl` after close has been called.
This feels like using `complete` field isn't the correct solution for this.

**Review by nicoloboschi — COMMENTED** [review]
Good analysis. 
I think, as said by @lhotari, we should use the compareAndSet. If we don't think it could happen to call `close()` twice, we can raise an exception if the comparison fails - it probably means there's another race uncovered

**Review by dlg99 — CHANGES_REQUESTED** [review]
1. `ListenerBasedPendingReadOp` is similar but does not close all entries in sequence on `submitCallback()`. I think we need to critically look at both. I didn't look at other ops.

2. Simply calling "close()" does not recycle entry's writeSet (memory leak). It's expected that the entry either fails or completes before close. This can be done either on close if the entry is not completed yet or, probably more explicit as

```
diff --git a/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java
index b48761ec6..18577a3de 100644
--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java
@@ -20,6 +20,7 @@
  */
 package org.apache.bookkeeper.client;
 
+import com.google.common.base.Preconditions;
 import com.google.common.collect.Lists;
 import com.google.common.util.concurrent.ListenableFuture;
 import io.netty.buffer.ByteBuf;
@@ -110,6 +111,7 @@ class PendingReadOp implements ReadEntryCallback, SafeRunnable {
 
         @Override
         public void close() {
+            Preconditions.checkArgument(isComplete(), "Closing non-completed entry");
             entryImpl.close();
         }
 
@@ -166,9 +168,15 @@ class PendingReadOp implements ReadEntryCallback, SafeRunnable {
          * @return true if we managed to fail the entry; otherwise return false if it already failed or completed.
          */
         boolean fail(int rc) {
+            return fail(rc, true);
+        }
+
+        boolean fail(int rc, boolean executeCallback) {
             if (complete.compareAndSet(false, true)) {
                 this.rc = rc;
-                submitCallback(rc);
+                if (executeCallback) {
+                    submitCallback(rc);
+                }
                 writeSet.recycle();
                 return true;
             } else {
@@ -641,7 +649,12 @@ class PendingReadOp implements ReadEntryCallback, SafeRunnable {
                     BKException.getMessage(code), firstUnread, firstRc);
             clientCtx.getClientStats().getReadOpLogger().registerFailedEvent(latencyNanos, TimeUnit.NANOSECONDS);
             // release the entries
-            seq.forEach(LedgerEntryRequest::close);
+            seq.forEach(entry -> {
+                // avoid calling the callback recursively
+                // fail() is re-entrant, won't do anything to already completed entries
+                entry.fail(code, false);
+                entry.close();
+            });
             future.completeExceptionally(BKException.create(code));
         } else {
             clientCtx.getClientStats().getReadOpLogger().registerSuccessfulEvent(latencyNanos, TimeUnit.NANOSECONDS);
```

**Comment by eolivelli** [comment]
Just to clarify: I don't think that not recycling a object is a memory leak, as the jvm will discard it, but we are losing the chanceto save a allocation.

BTW I second @dlg99 's comments

**Comment by congbobo184** [comment]
https://github.com/apache/bookkeeper/blob/e99cd4915b56ae3e9891ef9a83eb5e8f089505d7/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L146

complete will be change in this way and then close will not close ledgerEntryImpl if you use compareAndSet
@lhotari @nicoloboschi

**Comment by congbobo184** [comment]
@dlg99 hi, 
1. although `ListenerBasedPendingReadOp` don't close all ledgerEntryRequset, we only change complete, it will not change any logical for `ListenerBasedPendingReadOp`, we can fix `ListenerBasedPendingReadOp` in next pr, not in this PR
2. it will not memory leak when don't not recycle entry's writeSet. now in this PR we only fix LedgerEntryImpl.close.

**Comment by eolivelli** [comment]
I agree that we can follow up with a separate patch to ListenerBasedPendingReadOp

**Comment by lhotari** [comment]
> https://github.com/apache/bookkeeper/blob/e99cd4915b56ae3e9891ef9a83eb5e8f089505d7/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L146
> 
> complete will be changed in this way and then close will not close ledgerEntryImpl if you use compareAndSet
> 
> https://github.com/apache/bookkeeper/blob/e99cd4915b56ae3e9891ef9a83eb5e8f089505d7/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java#L169
> 
> if read fail, in this way we already change complete, so close will do anything. @lhotari @nicoloboschi

Yes I understand that, I posted a follow up comment https://github.com/apache/bookkeeper/pull/3110#issuecomment-1068136249 . Would you mind adding a code comment that explains the intent to future readers of the code?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java (hunk 2 above)
Comment (lhotari) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:113]: addressed by https://github.com/apache/bookkeeper/pull/3110#issuecomment-1068857090

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestParallelRead.java [hunk 3]
```diff
@@ -238,4 +247,59 @@ public void testParallelReadFailureWithFailedBookies() throws Exception {
...
+        assertTrue(first.complete.get());
+
+        assertEquals(-1, second.entryImpl.getEntryId());
+        assertEquals(-1, second.entryImpl.getLedgerId());
+        assertEquals(-1, second.entryImpl.getLength());
+        assertNull(second.entryImpl.getEntryBuffer());
+        assertTrue(second.complete.get());
+
+        // Mock ledgerEntryImpl reuse
+        Method method = PendingReadOp.class.getDeclaredMethod("createReadContext",
```
Comment (nicoloboschi) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestParallelRead.java:289]: what about making `ReadContext` class visibility to default and instantiate the object in the test ?

**Review by nicoloboschi — CHANGES_REQUESTED** [review]
I re-read the scenario and I now understand why the compare-and-set would not work.
Good work!

**Review by lhotari — APPROVED** [review]

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestParallelRead.java (hunk 3 above)
Comment (congbobo184) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/TestParallelRead.java:289]: This PR is best not to change the ReadContext class visibility, keep the original level

**Comment by congbobo184** [comment]
rerun failure checks

**Comment by eolivelli** [comment]
@congbobo184 what about applying @dlg99 's patch to your branch ?

**Comment by congbobo184** [comment]
@dlg99 I have add writeSet recycle logical, please review again, thanks

**Comment by congbobo184** [comment]
rerun failure checks

**Comment by congbobo184** [comment]
rerun failure checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java (hunk 2 above)
Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:119]: Now that we have the `writeSet.recycle();` block above we could go back to compareAndSet()

What about this ?

```
if (complete.compareAndSet(false, true)) {
   writeSet.recycle();
}
```

This way:
- we do writeSet.recycle(); only if complete is "false"
- eventually complete will be "true" in any case

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java [hunk 4]
```diff
@@ -580,18 +590,23 @@ void sendReadTo(int bookieIndex, BookieId to, LedgerEntryRequest entry) throws I
...
     @Override
     public void readEntryComplete(int rc, long ledgerId, final long entryId, final ByteBuf buffer, Object ctx) {
+
         final ReadContext rctx = (ReadContext) ctx;
         final LedgerEntryRequest entry = rctx.entry;
 
         if (rc != BKException.Code.OK) {
-            entry.logErrorAndReattemptRead(rctx.bookieIndex, rctx.to, "Error: " + BKException.getMessage(rc), rc);
+            // if entry complete = true, don't need logErrorAndReattemptRead
+            if (!entry.complete.get()) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:599]: we should add a test covering this change: calling readEntryComplete with a non-OK error code twice

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java [hunk 5]
```diff
@@ -580,18 +590,23 @@ void sendReadTo(int bookieIndex, BookieId to, LedgerEntryRequest entry) throws I
...
+                entry.logErrorAndReattemptRead(rctx.bookieIndex, rctx.to, "Error: " + BKException.getMessage(rc), rc);
+            }
             return;
         }
 
         heardFromHosts.add(rctx.to);
         heardFromHostsBitSet.set(rctx.bookieIndex, true);
 
         buffer.retain();
+        // if entry has completed don't handle twice
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:609]: do we have a test for this case ?
can you add this case to your new tests please ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java (hunk 2 above)
Comment (congbobo184) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:119]:
```
            if (!complete.get()) {
                writeSet.recycle();
            }
            complete.set(true);
```
I think that would also work

**Comment by congbobo184** [comment]
rerun failure checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java (hunk 2 above)
Comment (dlg99) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:119]: @congbobo184 with this approach we risk calling recycle twice if close() is called from different threads. 
At least we need to implement @eolivelli 's suggestion. I'd set rc to something as well, eg:

```
if (complete.compareAndSet(false, true)) {
   rc = BKException.Code.UnexpectedConditionException;
   writeSet.recycle();
}
```

**Comment by congbobo184** [comment]
rerun failure checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java (hunk 2 above)
Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:119]: @congbobo184 this is the last comment, then we are good to go (and also cut a release)

**Comment by congbobo184** [comment]
rerun failure checks

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java [hunk 6]
```diff
@@ -110,6 +110,14 @@
...
         @Override
         public void close() {
+            // this request has succeeded before, can't recycle writeSet again
+            if (complete.compareAndSet(false, true)) {
+                rc = BKException.Code.UnexpectedConditionException;
+                writeSet.recycle();
+            }
+            // To be able to close this when complete = true, can't use compareAndSet.
+            // Because readComplete will make complete = true then close will not close LedgerEntryImpl
+            complete.set(true);
```
Comment (dlg99) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingReadOp.java:120]: this is not needed anymore

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by dlg99 — APPROVED** [review]
Looks great!
Thank you.

**Comment by congbobo184** [comment]
rerun failure checks

**Pull request merged**